device/iot-device-samples/send-event/src/main/java/samples/com/microsoft/azure/sdk/iot/SendEvent.java

```diff
@@ -10,233 +10,57 @@
-            }
-
-            if (args.length == 3)
-            {
-                pathToCertificate = null;
-            }
-            else
-            {
-                pathToCertificate = args[3];
-            }
-        }
-
-
-        System.out.println("Successfully read input parameters.");
-        System.out.format("Using communication protocol %s.\n", protocol.name());
-
-        DeviceClient client = new DeviceClient(connString, protocol);
-        if (pathToCertificate != null )
-        {
-            client.setOption("SetCertificatePath", pathToCertificate );
-        }
-
-        System.out.println("Successfully created an IoT Hub client.");
-
-        // Set your token expiry time limit here
-        long time = 2400;
-        client.setOption("SetSASTokenExpiryTime", time);
-        System.out.println("Updated token expiry time to " + time);
-
-        client.registerConnectionStatusChangeCallback(new IotHubConnectionStatusChangeCallbackLogger(), new Object());
+        final long messagesToSend = 1000000;
+        String connString = "<device connection string for a device in your B3 or S3 iot hub>";
+        DeviceClient client = new DeviceClient(connString, IotHubClientProtocol.MQTT);
+        long clientSendInterval = 10; //Lower number here spawns send threads more frequently, can send more quickly. By default, value is 10
+        client.setOption("SetSendInterval", clientSendInterval);
+        countDownLatch = new CountDownLatch((int) messagesToSend);
 
         client.open();
 
-        System.out.println("Opened connection to IoT Hub.");
-        System.out.println("Sending the following event messages:");
-
-        String deviceId = "MyJavaDevice";
-        double temperature = 0.0;
-        double humidity = 0.0;
-
-        for (int i = 0; i < numRequests; ++i)
+        final long startTime = System.currentTimeMillis();
+        while (sentMessageCount < messagesToSend)
         {
-            temperature = 20 + Math.random() * 10;
-            humidity = 30 + Math.random() * 20;
-
-            String msgStr = "{\"deviceId\":\"" + deviceId +"\",\"messageId\":" + i + ",\"temperature\":"+ temperature +",\"humidity\":"+ humidity +"}";
-
-            try
-            {
-                Message msg = new Message(msgStr);
-                msg.setContentType("application/json");
-                msg.setProperty("temperatureAlert", temperature > 28 ? "true" : "false");
-                msg.setMessageId(java.util.UUID.randomUUID().toString());
-                msg.setExpiryTime(D2C_MESSAGE_TIMEOUT);
-                System.out.println(msgStr);
-
-                EventCallback callback = new EventCallback();
-                client.sendEventAsync(msg, callback, msg);
-            }
-
-            catch (Exception e)
-            {
-                e.printStackTrace(); // Trace the exception
-            }
+            Message message = new Message("a");
+            EventCallback callback = new EventCallback();
```

Comment:
And teh right way to measure time here would be to pass the start time in callback and note the stop time when ack is received. If not you are accounting for creating of the message as well. Even if that does't cause network IO, it may change the metrics.

Reply:
I can move the message creation outside of the timed loop, instead. I'll just have an array of messages created for the timed loop to pull from. 

I don't follow putting the start time in the callback. We are measuring the time it takes to send thousands of messages, not the time it takes to send each message